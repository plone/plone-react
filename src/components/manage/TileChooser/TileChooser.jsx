import React from 'react';
import PropTypes from 'prop-types';
import { filter, map, groupBy } from 'lodash';
import { Accordion, Button } from 'semantic-ui-react';
import { injectIntl } from 'react-intl';
import { Icon } from '@plone/volto/components';
import AnimateHeight from 'react-animate-height';
import { tiles } from '~/config';

import upSVG from '@plone/volto/icons/up-key.svg';
import downSVG from '@plone/volto/icons/down-key.svg';

const TileChooser = ({ currentTile, onMutateTile, intl }) => {
  const mostUsedTiles = filter(tiles.tilesConfig, item => item.mostUsed);
  const groupedTiles = groupBy(tiles.tilesConfig, item => item.group);
  const tilesAvailable = { mostUsed: mostUsedTiles, ...groupedTiles };
  const [activeIndex, setActiveIndex] = React.useState(0);

  function handleClick(e, titleProps) {
    const { index } = titleProps;
    const newIndex = activeIndex === index ? -1 : index;

    setActiveIndex(newIndex);
  }

  return (
    <div className="tiles-chooser">
      <Accordion fluid styled className="form">
        {map(tiles.groupTilesOrder, (groupName, index) => (
          <React.Fragment key={groupName.id}>
            <Accordion.Title
              active={activeIndex === index}
              index={index}
              onClick={handleClick}
            >
              {intl.formatMessage({
                id: groupName.id,
                defaultMessage: groupName.title,
              })}
              <div className="accordion-tools">
                {activeIndex === 0 ? (
                  <Icon name={upSVG} size="20px" />
                ) : (
                  <Icon name={downSVG} size="20px" />
                )}
              </div>
            </Accordion.Title>
            <Accordion.Content
              className={groupName.id}
              active={activeIndex === index}
            >
              <AnimateHeight
                animateOpacity
                duration={500}
                height={activeIndex === index ? 'auto' : 0}
              >
                {map(
                  filter(
                    tilesAvailable[groupName.id],
                    tile => !tile.restricted,
                  ),
                  tile => (
                    <Button.Group key={tile.id}>
                      <Button
                        icon
                        basic
                        className={tile.id}
                        onClick={() =>
                          onMutateTile(currentTile, { '@type': tile.id })
                        }
                      >
                        <Icon name={tile.icon} size="36px" />
                        {intl.formatMessage({
                          id: tile.id,
                          defaultMessage: tile.title,
                        })}
                      </Button>
                    </Button.Group>
                  ),
                )}
              </AnimateHeight>
            </Accordion.Content>
          </React.Fragment>
        ))}
      </Accordion>
    </div>
  );
};

TileChooser.propTypes = {
  currentTile: PropTypes.string.isRequired,
  onMutateTile: PropTypes.func.isRequired,
};

export default injectIntl(TileChooser);
