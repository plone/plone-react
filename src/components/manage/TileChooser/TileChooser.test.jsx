import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-intl-redux';
import configureStore from 'redux-mock-store';
import TileChooser from './TileChooser';

const mockStore = configureStore();

test('renders a TileChooser component', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <TileChooser onMutateTile={() => {}} currentTile="thetileid" />
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
