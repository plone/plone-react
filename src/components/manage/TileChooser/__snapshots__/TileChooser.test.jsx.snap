// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a TileChooser component 1`] = `
<div
  className="tiles-chooser"
>
  <div
    className="accordion ui fluid styled form"
  >
    <div
      className="active title"
      onClick={[Function]}
    >
      Most used
      <div
        className="accordion-tools"
      >
        <svg
          className="icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          onClick={null}
          style={
            Object {
              "fill": "currentColor",
              "height": "20px",
              "width": "auto",
            }
          }
          viewBox=""
          xmlns=""
        />
      </div>
    </div>
    <div
      className="content active tiles-list"
    >
      <div
        aria-hidden={false}
        className="rah-static rah-static--height-auto"
        style={
          Object {
            "height": "auto",
            "overflow": "visible",
          }
        }
      >
        <div
          style={
            Object {
              "WebkitTransition": "opacity 500ms ease 0ms",
              "transition": "opacity 500ms ease 0ms",
            }
          }
        >
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Image
            </button>
          </div>
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Video
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="title"
      onClick={[Function]}
    >
      Text
      <div
        className="accordion-tools"
      >
        <svg
          className="icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          onClick={null}
          style={
            Object {
              "fill": "currentColor",
              "height": "20px",
              "width": "auto",
            }
          }
          viewBox=""
          xmlns=""
        />
      </div>
    </div>
    <div
      className="content tiles-list"
    >
      <div
        aria-hidden={true}
        className="rah-static rah-static--height-zero"
        style={
          Object {
            "height": 0,
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "WebkitTransition": "opacity 500ms ease 0ms",
              "opacity": 0,
              "transition": "opacity 500ms ease 0ms",
            }
          }
        >
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Text
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="title"
      onClick={[Function]}
    >
      Media
      <div
        className="accordion-tools"
      >
        <svg
          className="icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          onClick={null}
          style={
            Object {
              "fill": "currentColor",
              "height": "20px",
              "width": "auto",
            }
          }
          viewBox=""
          xmlns=""
        />
      </div>
    </div>
    <div
      className="content tiles-list"
    >
      <div
        aria-hidden={true}
        className="rah-static rah-static--height-zero"
        style={
          Object {
            "height": 0,
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "WebkitTransition": "opacity 500ms ease 0ms",
              "opacity": 0,
              "transition": "opacity 500ms ease 0ms",
            }
          }
        >
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Image
            </button>
          </div>
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Video
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="title"
      onClick={[Function]}
    >
      Common
      <div
        className="accordion-tools"
      >
        <svg
          className="icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          onClick={null}
          style={
            Object {
              "fill": "currentColor",
              "height": "20px",
              "width": "auto",
            }
          }
          viewBox=""
          xmlns=""
        />
      </div>
    </div>
    <div
      className="content tiles-list"
    >
      <div
        aria-hidden={true}
        className="rah-static rah-static--height-zero"
        style={
          Object {
            "height": 0,
            "overflow": "hidden",
          }
        }
      >
        <div
          style={
            Object {
              "WebkitTransition": "opacity 500ms ease 0ms",
              "opacity": 0,
              "transition": "opacity 500ms ease 0ms",
            }
          }
        >
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Hero
            </button>
          </div>
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Maps
            </button>
          </div>
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              HTML
            </button>
          </div>
          <div
            className="ui buttons"
          >
            <button
              className="ui basic icon button"
              onClick={[Function]}
            >
              <svg
                className="icon"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": undefined,
                  }
                }
                onClick={null}
                style={
                  Object {
                    "fill": "currentColor",
                    "height": "36px",
                    "width": "auto",
                  }
                }
                viewBox=""
                xmlns=""
              />
              Table
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
